microsoft/PowerToys
# [FancyZones] Monitor identification

## Summary of the Pull Request

Changed monitor (both hardware and virtual) monitor identifying for better stability. 

```
"device": {
   "monitor": "TLX1388",
   "monitor-instance": "4&125707d6&0&UID8388688",
   "serial-number": "TL139GDXP03",
   "virtual-desktop": "{A055482B-3629-480C-992F-4B31BF9A9771}"
}
```


## PR Checklist

- [x] **Closes:** #16370


## Detailed Description of the Pull Request / Additional comments


## Validation Steps Performed

1. Check update from the previous version
    * On the previous version set any layout and snap a window to a zone
    * Run FancyZones
    * Verify that layout is the same and window is opening in the correct zone (with `Move newly created windows to their last known zone` enabled)
2. Check layout reset
2.1 Without virtual desktop id
    * Remove virtual desktops from the registry (`` and ``)
    * Set any layout different from the default on each monitor
    * Snap a window to a zone
    * Restart/lock computer
    * Verify you have the same layout and window is opening in the right place 
2.1 With virtual desktop id
    * Add a virtual desktop
    * Set any layout different from the default on each monitor on both virtual desktops
    * Snap a window to a zone
    * Restart/lock computer
    * Verify you have the same layout and window is opening in the right place
4. Check layout reset on VM/RDP
    * Set any layout and snap a window to a zone
    * Restart/lock
    * Verify layout and window position
5. Check `Allow zones to span across monitors` behavior
    * Set any layout on each monitor
    * Enable `Allow zones to span across monitors`
    * Set any layout on combined monitors
    * Disable `Allow zones to span across monitors`
    * Verify separate layouts weren't reset
    * Enable `Allow zones to span across monitors` back
    * Verify combined layout wasn't reset
6. Check update from an older version
    * Create `zones-settings.json` in the `\AppData\Local\Microsoft\PowerToys\FancyZones`. The example is below, you can change `device-id` to fit your actual monitor. Schema is `monitor-id_width_height_{virtual-desktop-id}`.
    * Update `app-zone-history.json` in the same folder. Update `device-id` to match one from the `zones-settings.json`
    * Start FancyZones
    * Verify layout wasn't reset and the app from `app-zone-history` is opening in the target zone
    
**zones-settings.json**
```
{
    "devices": [
        {
            "device-id": "TLX1388#4&125707d6&0&UID8388688_3000_2000_{5258AF51-D761-4B3A-8DF8-EA16359E71D1}",
            "active-zoneset": {
                "uuid": "{0120E707-CA34-4FD1-B647-2EE344A7415A}",
                "type": "custom"
            },
            "editor-show-spacing": false,
            "editor-spacing": 0,
            "editor-zone-count": 22,
            "editor-sensitivity-radius": 20
        }
    ],
    "custom-zone-sets": [
        {
            "uuid": "{0120E707-CA34-4FD1-B647-2EE344A7415A}",
            "name": "CustomGrid",
            "type": "grid",
            "info": {
                "rows": 2,
                "columns": 3,
                "rows-percentage": [
                    5000,
                    5000
                ],
                "columns-percentage": [
                    5017,
                    3313,
                    1670
                ],
                "cell-child-map": [
                    [
                        0,
                        1,
                        2
                    ],
                    [
                        0,
                        1,
                        3
                    ]
                ],
                "sensitivity-radius": 20,
                "show-spacing": true,
                "spacing": 16
            }
        }
    ],
    "templates": [
        {
            "type": "blank",
            "show-spacing": false,
            "spacing": 0,
            "zone-count": 0,
            "sensitivity-radius": 0
        },
        {
            "type": "focus",
            "show-spacing": false,
            "spacing": 0,
            "zone-count": 3,
            "sensitivity-radius": 20
        },
        {
            "type": "columns",
            "show-spacing": true,
            "spacing": 16,
            "zone-count": 3,
            "sensitivity-radius": 20
        },
        {
            "type": "rows",
            "show-spacing": true,
            "spacing": 16,
            "zone-count": 3,
            "sensitivity-radius": 20
        },
        {
            "type": "grid",
            "show-spacing": true,
            "spacing": 16,
            "zone-count": 3,
            "sensitivity-radius": 20
        },
        {
            "type": "priority-grid",
            "show-spacing": true,
            "spacing": 16,
            "zone-count": 3,
            "sensitivity-radius": 20
        }
    ],
    "quick-layout-keys": []
}
```

**app-zone-history.json**
```
{
    "app-zone-history": [
        {
            "app-path": "...\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe",
            "history": [
                {
                    "zone-index-set": [
                        1
                    ],
                    "device-id": "TLX1388#4&125707d6&0&UID8388688_3000_2000_{5258AF51-D761-4B3A-8DF8-EA16359E71D1}",
                    "zoneset-uuid": "{0120E707-CA34-4FD1-B647-2EE344A7415A}"
                }
            ]
        }
    ]
}
```
